### DISCUSSION TITLE
[@types/responselike] Cannot find type definition file for 'responselike'

### DISCUSSION THREAD

**Discussion by mfazekas** [DefinitelyTyped/DefinitelyTyped#62111] [Issues with a @types package]
When running our tsc commands in CI we're now getting.
```
error TS2688: Cannot find type definition file for 'responselike'.
  The file is in the program because:
    Entry point for implicit type library 'responselike'


Found 1 error.
```

It seems that `@types/responselike`/3.0.0 was recently deprecated and it doesn't contains index.d.ts, which tsc doesn't like.

I could work around by adding 

```json
 "optionalDependencies": {
    "@types/responselike": "1.0.0"
}
```

Not sure if there is a better workaround, solution, for this issue.

Our `package.json`

```json
...
  "dependencies": {
    "@mapbox/geo-viewport": "^0.5.0",
    "@mapbox/mapbox-sdk": "^0.13.4",
...
```

Our `yarn.lock` looks like:
```lock

"@mapbox/mapbox-sdk@^0.13.4":
  version "0.13.4"
  resolved "https://registry.yarnpkg.com/@mapbox/mapbox-sdk/-/mapbox-sdk-0.13.4.tgz#0d3292f325355a520babdc390a348103695d2f12"
  integrity sha512-8ecFC+eWLe0/6FKQuhhSs5cFgvvc07In2r9KdY3h8tP9weH0Fzh7B5Zjj8fUFzAHi8pFQ5KguEYfqrnPJN/PdQ==
  dependencies:
    "@mapbox/fusspot" "^0.4.0"
    "@mapbox/parse-mapbox-token" "^0.2.0"
    "@mapbox/polyline" "^1.0.0"
    eventemitter3 "^3.1.0"
    form-data "^3.0.0"
    got "^10.7.0"

...

got@^10.7.0:
  version "10.7.0"
  resolved "https://registry.yarnpkg.com/got/-/got-10.7.0.tgz#62889dbcd6cca32cd6a154cc2d0c6895121d091f"
  integrity sha512-aWTDeNw9g+XqEZNcTjMMZSy7B7yE9toWOFYip7ofFTLleJhvZwUxxTxkTpKvF+p1SAA4VHmuEy7PiHTHyq8tJg==
  dependencies:
    "@sindresorhus/is" "^2.0.0"
    "@szmarczak/http-timer" "^4.0.0"
    "@types/cacheable-request" "^6.0.1"
    cacheable-lookup "^2.0.0"

...

"@types/cacheable-request@^6.0.1":
  version "6.0.2"
  resolved "https://registry.yarnpkg.com/@types/cacheable-request/-/cacheable-request-6.0.2.tgz#c324da0197de0a98a2312156536ae262429ff6b9"
  integrity sha512-B3xVo+dlKM6nnKTcmm5ZtY/OL8bOAOd2Olee9M1zft65ox50OzjEHW91sDiU9j6cvW8Ejg1/Qkf4xd2kugApUA==
  dependencies:
    "@types/http-cache-semantics" "*"
    "@types/keyv" "*"
    "@types/node" "*"
    "@types/responselike" "*"
    
...


"@types/responselike@*":
  version "3.0.0"
  resolved "https://registry.yarnpkg.com/@types/responselike/-/responselike-3.0.0.tgz#5ecc1fc88552e5ac03de648a7796f9e125d002dc"
  integrity sha512-zfgGLWx5IQOTJgQPD4UfGEhapTKUPC1ra/QCG02y3GUJWrhX05bBf/EfTh3aFj2DKi7cLo+cipXLNclD27tQXQ==
  dependencies:
    responselike "*"
```

[a comment by an automated account is not shown]

**Comment by BendingBender** [comment]
@mfazekas This works as expected. If you use an older version of `responselike` you still need to install the older version of the `@types/responselike` package. In this case it's `@types/responselike@1.0.0`. You can install it via:
```
npm i -D @types/responselike@1.0.0
```

It's unfortunate that `@types/responselike` is listed with a wildcard in `@types/cacheable-request` but this is how DTS currently works. You can open an issue in this repo.

**Reply by BendingBender** [reply]
I've asked on the TS discord what we can do to fix this: https://discord.com/channels/508357248330760243/757992230077333655

**Comment by inoue2002** [comment]
Hi, I was just struggling with the same error, but I solved it by installing the package as follows🎉 Thanks :)

```
npm install @types/responselike@1.0.0 --save-dev
```

**Comment by garethstaite** [comment]
We are now running into issues due to the fact we are still running node 12 for some projects.
Everything that depends on `responselike@3.0.0` via the wild card [will not build unless you are running node >=14](https://github.com/sindresorhus/responselike/blob/main/package.json#L12).

We do not control this dependency tree (`got` > `@types/cacheable-request` > `@types/responselike` > `responselike`).

I can only assume before the [latest changes](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/c854edb2f4c9c80a9b0511e3269b9bd92c2eccdf) the tree would have looked like:
- [`got@11.8.5` depending on `@types/cacheable-request@6.0.1`](https://github.com/sindresorhus/got/blob/v11.8.5/package.json#L48) (resolved to 6.0.2)
- `@types/cacheable-request@6.0.2` depending on `@types/responselike@*` (resolved to 1.0.0)
- `@types/responselike@1.0.0` does not depend on `responselike` (it only depends on `@types/node@*`)

And now the tree looks like:
- [`got@11.8.5` depending on `@types/cacheable-request@6.0.1`](https://github.com/sindresorhus/got/blob/v11.8.5/package.json#L48) (resolved to 6.0.2)
- `@types/cacheable-request@6.0.2` depending on `@types/responselike@*` (resolved to 3.0.0)
- `@types/responselike@3.0.0` depending on `responselike@*` resolved to 3.0.0
- `responselike@3.0.0` does not support node 12

**Reply by BendingBender** [reply]
You can temporarily fix this by explicitly installing:
```
npm i -D @types/cacheable-request@6.0.2 @types/responselike@1.0.0
```

**Comment by BendingBender** [comment]
Ok, the DTS maintainers now have moved the `latest` tag back to `@types/responselike@1.0.0`. Plz let me know if that fixes your problems.

**Comment by nbouvrette** [comment]
None of these workarounds worked for me. Until the [root of the issue](https://github.com/sindresorhus/got/issues/2138) is resolved, this one did the trick: (requires npm [v8.3.0](https://github.com/npm/cli/releases/tag/v8.3.0) (2021-12-09)):

Update your `package.json` and add:

```json
  "overrides": {
    "@types/cacheable-request": "8.3.1"
  },
```

Edit `/node_modules/cacheable-request/dist/types.d.ts` and remove the 4 first lines `/// <reference types="node" resolution-mode="require"/>` (they are triggering a `'resolution-mode' assertions are only supported when 'moduleResolution' is 'node16' or 'nodenext'.` error and I could not find anything on Google... removing them worked for me)
